feature/google/src/latestDependencies/java/com/revenuecat/purchases/google/BillingWrapper.kt

```diff
@@ -765,4 +760,28 @@ class BillingWrapper(
             listener.onPurchaseHistoryResponse(devErrorResponseCode, null)
         }
     }
+
+    private fun createPurchaseParams(
+        storeProduct: StoreProduct,
+        purchaseOption: PurchaseOption
+    ): BillingFlowParams? {
+        val token = purchaseOption.token
+        val productDetails = storeProduct.productDetails
+        if (token == null) {
+            errorLog("PurchaseOption must have a token with BC5.") //TODOBC5: Improve and move error message
+            return null
+        }
+        if (productDetails == null) {
```

Comment:
I think here we should be downcasting to GooglePurchaseOption to access purchaseDetails. If we don't do that we'll be again:
a) exposing Google-specific details in the public interface that aren't relevant
b) requiring Amazon to implement this for the PurchaseOption which doesn't make sense

Reply:
So the `StoreProduct.productDetails` extension is doing the downcast to `GoogleStoreProduct` + accesing the `productDetails` there directly.

> a) exposing Google-specific details in the public interface that aren't relevant

I think we should first list what we want to expose exactly. Personally I feel exposing the `GoogleStoreProduct` (for either BC4 and BC5) and `AmazonStoreProduct` is not the best option since they seem to be bigger data classes that will bloat the public API and they are also exposing store specific classes (BC5 `GoogleStoreProduct` exposes `ProductDetails`, BC4 `GoogleStoreProduct` exposes `SkuDetails`, `AmazonStoreProduct` exposes Amazon's `Product` class). After we've decided what to expose, we can add that to the API tester project to make sure there are no regressions there.

> b) requiring Amazon to implement this for the PurchaseOption which doesn't make sense

The way the extension is implemented, if we try to call `StoreProduct.productDetails` with an amazon StoreProduct, it would return null. We could remove the extension and just do the downcast here but we would have to guard against the possibility of the product not being a `GoogleStoreProduct` and then in the amazon code, we would have to also do the downcast and guard against the possibility of the product not being an AmazonStoreProduct. I guess it's just moving some logic from an extension to this place but with an extension is a bit more reusable.

Reply:
a) If we're trying to provide a uniform interface that doesn't require the user to know specific let's just expose: a StoreProduct (title, duration, etc...) with a list of PurchaseOffer(pricingPhases). All details of whether this is backed by a ProductDetails, whether it uses offerTokens or something else, can be in implementations. You should be able to build a paywall and buying something without touching anything Amazon or Google specific, right? We don't need to expose any of these implementation details to make the API work

b)  I don't like this approach much. It ends up overloading the API with all options. Will the Amazon one have the amazonProduct? it means we have a class with a lot of implementation specific members, and for a specific implementation all except one will be null. IMO that's what subclasses are for.

Reply:
Ok, I thought we were planning to expose the original products from the stores from some comments https://github.com/RevenueCat/purchases-android/pull/631#discussion_r994654473 and https://github.com/RevenueCat/purchases-android/pull/658#discussion_r1006593644.

In that case, should the original products be removed from the `GoogleStoreProduct` and `AmazonStoreProduct` classes? Otherwise we are still leaking store SDK classes, we are just adding a level of indirection.